dafni_cli/api/session.py

```diff
@@ -327,15 +348,88 @@ def _authenticated_request(
+
+        Args:
+            response (requests.Response): The failed request response
+
+        Returns:
+            Optional[str]: String representing an error message or None
+                           if none was found
+        """
+
+        # Try and get JSON data from the response
+        try:
+            error_message = None
+            decoded_response = response.json()
+            # Some requests have an error and error_message, in which case we
+            # want to override with the latter
+            if "error" in decoded_response:
+                error_message = f"Error: {decoded_response['error']}"
+            if "error_message" in decoded_response:
+                error_message = f"{error_message}, {decoded_response['error_message']}"
+            elif "errors" in decoded_response:
+                error_message = "The following errors were returned:"
+                for error in decoded_response["errors"]:
+                    error_message += f"\nError: {error}"
+            # Special case when uploading dataset metadata that's invalid
+            elif "metadata" in decoded_response:
+                # This returns a list of errors, add them all to the
+                # message
+                error_message = "Found errors in metadata:"
+                for error in decoded_response["metadata"]:
+                    error_message += f"\n{error}"
+
+            return error_message
+        except requests.JSONDecodeError:
+            return None
+
+    def _check_response(self, url: str, response: requests.Response):
+        """Checks a requests response for any errors and raises them as
+        required
+
+        Args:
+            url (str): URL endpoint that was being queried
+            response (requests.Response): Response from requests
+
+        Raises:
+            EndpointNotFoundError: If the response returns a 404 status code
+            DAFNIError: If an error occurs with an error message from DAFNI
+            HTTPError: If any other error occurs without an error message from
+                       DAFNI
+        """
+
+        error_message = None
+
+        # Check for any error response
+        if not response.ok:
+            # Specialised error for when we get a 404 - helps to identify
+            # missing objects
+            if response.status_code == 404:
+                raise EndpointNotFoundError(f"Could not find {url}")
+
+            # Attempt to find an error message from the API itself
+            error_message = self.get_error_message(response)
+
+        # If there is an error from DAFNI raise a DAFNI exception as well
+        # with more details, otherwise leave as any errors are HTTPError's
+        if error_message is None:
+            response.raise_for_status()
+        else:
+            try:
+                response.raise_for_status()
+            except HTTPError as err:
+                raise DAFNIError(error_message) from err
```

Comment:
Could reduce nesting a little by doing 
```suggestion
        try:
            response.raise_for_status()
        except HTTPError as err:
            if error_message is None:
                raise err
            raise DAFNIError(error_message) from err
```

Reply:
Thanks!